### ISSUE TITLE
Chrome version error

### ISSUE THREAD

**Issue by fourstacks** [orchestral/testbench-dusk#37]
Hello, thanks for all your work on this package.  I've just tried to get things up and running to test a package I'm writing and I've hit an error that looks like it might be related to: https://github.com/laravel/dusk/issues/641

Basically I have a super stripped-back test just to get things up and running and it's throwing the following error:

```
Facebook\WebDriver\Exception\SessionNotCreatedException: session not created: 
Chrome version must be between 70 and 73 
(Driver info: chromedriver=2.45.615355 
(d5698f682d8b2742017df6c81e0bd8e6a3063189),
platform=Mac OS X 10.14.0 x86_64)
```

(dropped some line spaces in here to make things more readable but it's otherwise the error I get in terminal).

I believe my test itself is probably fine (it's in this gist for sanity checking purposes - I've switched up the namespace to something generic - https://gist.github.com/fourstacks/09f5fa1d9f2a1b8eb45ef1189413850f)

As I'm unable to run `php artisan dusk:chrome-driver` (due to this being a package and artisan not being available) I'm at a bit of a loss to know what else is possible.

If anyone else has run into the same issue when using testbench-dusk I'd really appreciate any tips or pointers.  Thanks!

**Comment by Keoghan** [comment]
Hi, 

Thanks for using the package and for your feedback. At first look, this does appear to be related to the issue you have highlighted.  

We might need to perform a check when running the tests to validate the chrome-driver and run the dusk:chrome-driver command if it is not compatible. Or look at providing a way to run it manually. (Probably both).

Open to thoughts, will see what I can come up with :)

Thanks, Keoghan






On 7 May 2019, at 20:11, John Wyles <notifications@github.com> wrote:

Hello, thanks for all your work on this package. I've just tried to get things up and running to test a package I'm writing and I've hit an error that looks like it might be related to: laravel/dusk#641

Basically I have a super stripped-back test just to get things up and running and it's throwing the following error:

Facebook\WebDriver\Exception\SessionNotCreatedException: session not created: Chrome version must be between 70 and 73 (Driver info: chromedriver=2.45.615355 (d5698f682d8b2742017df6c81e0bd8e6a3063189),platform=Mac OS X 10.14.0 x86_64)
I believe my test itself is probably fine (it's in this gist for sanity checking purposes - I've switched up the namespace to something generic - https://gist.github.com/fourstacks/09f5fa1d9f2a1b8eb45ef1189413850f)

As I'm unable to run php artisan dusk:chrome-driver (due to this being a package and artisan not being available) I'm at a bit of a loss to know what else is possible.

If anyone else has run into the same issue when using testbench-dusk I'd really appreciate any tips or pointers. Thanks!

**Comment by Keoghan** [comment]
I've put together a basic version where the chrome-driver can be updated manually (tested against my development version of testbench dusk which failed with the same error).

https://github.com/orchestral/testbench-dusk/tree/feature/add-chromedriver-command

Please let me know your thoughts/mileage, then perhaps we can add some automated check/update when running tests.

**Comment by fourstacks** [comment]
Thanks for such a speedy response!  So I pulled in that branch but I'm afraid I've not had much joy so far.

I ran `php vendor/orchestra/testbench-dusk/artisan dusk:chrome-driver` as opposed to the command you mentioned in the post above (which spat out that the artisan file could not be found which makes sense).

However, running that command above results in a PHP error.  Basically in the artisan file that's now been added to the package an `autoload.php` is referenced within  `__DIR__.'/vendor/autoload.php';` on line 18.  Because this file does not exist at this path PHP throws a `failed to open stream: No such file or directory` error.

I can get around that by changing line 18 to:

`require 'vendor/autoload.php';`

To reference the autoload file at the root of the packages `vendor` folder.  (Not sure if that's really a good way of handling this though).

However, this then results in basically the same kind of error on line 21 where PHP cannot find the `app.php` file at `'vendor/laravel/laravel/bootstrap/app.php';`

This one is harder to figure out though.  The `app.php` does not exist in the `testbench-dusk/laravel/bootstrap` directory.

Been using Laravel for ages though not familiar enough with the internals of dusk and the orchestra testbench package to have a good suggestion here.  Happy to answer any followup Qs.

**Comment by Keoghan** [comment]
Ah ha, sorry about that - I'll pop into a package to resolve.  I had overlooked the different autoload/file locations.

**Comment by Keoghan** [comment]
Hi @fourstacks - I've just updated the branch. Hopefully the `php vendor/orchestra/testbench-dusk/artisan dusk:chrome-driver` will work for you now.

**Comment by fourstacks** [comment]
Nice - that worked a treat!  Thanks very much for the speedy response to this - much appreciated.

Any idea how often this issue is likely to crop up?  If it's something that's likely to be a reoccurring issue then some sort of pre-run check that runs that command to update the driver would probably be a worthwhile investment - hopefully reduce the chance of someone posting a similar issue in 6 months time...

**Comment by Keoghan** [comment]
I think a pre-flight check will be handy too, will see what's possible with that. At a minimum a nicer error message and suggested action would be beneficial.

**Comment by crynobone** [comment]
@Keoghan I'm wondering if the command can be built as a standalone Symfony command, If say you want to use Laravel Dusk on standalone usage, a Laravel skeleton isn't always available.

**Comment by Keoghan** [comment]
@crynobone - I don't quite follow, sorry...

Are you thinking that Testbench Dusk would be used for the standalone testing? If using Testbench Dusk, we can rely on the Laravel skeleton in the package - as it's only really used to boot up and provide access to the command.

Or using Dusk directly (and so the Symfony command would need to be in Dusk)?

Or, a standalone package to provide just that Symfony command, so it can be pulled in elsewhere?

**Comment by crynobone** [comment]
Are you thinking that Testbench Dusk would be used for the standalone testing?

Standalone here means to run a request to external URL such as `http://google.com` and automate task. I have one project testing my production app using Laravel Dusk and getting the same `ChromeDriver` error. 

It should be similar to <https://github.com/konsulting/dusk-standalone>

-------

So the suggestion for the chrome updater to be available as a symfony command is that we can pull this separated package and run it via `vendor/bin/dusk-chrome-updater` and it would work without having to bootstrap Laravel.

**Comment by crynobone** [comment]
@Keoghan 

Nevermind, I just created <https://github.com/orchestral/dusk-updater> for this purpose. However it would be good if we can provide some suggested `.travis.yml` template for Testbench Dusk.

**Comment by Keoghan** [comment]
@crynobone 👍 will take a look at Travis template

**Comment by crynobone** [comment]
Released `v3.6.2`, `v3.7.1` and `v3.8.1`
